new rule: HTML page language subtag matches default language (#1334)
act-rules/act-rules.github.io: pages/glossary/default-page-language.md

---
title: Default Page Language
key: default-page-language
unambiguous: true
objective: true
input_aspects:
  - DOM tree
  - Accessibility tree
  - CSS Styling
  - Language
---

The default page language is the _most common language_ used in a [web page][]. If there are no _words_ in the page, the default language is undefined.

> The _most common language_ is determined by counting the number of _words_ in the [web page][] that are part of any of the languages in the [language subtag registry][]. The same word can be part of multiple languages. If the number of words is the same, the default language is undefined.

Reviewer: I'm afraid this requires dictionaries (as in, "list of words") in addition to the languages.

Different dictionaries, even for the same language, do not list the exact same words, notably for new words.
For example, as of now, the French dictionary Larousse does not accept "infox" while Le Petit Robert does accept it.
https://www.larousse.fr/dictionnaires/francais/infox
https://dictionnaire.lerobert.com/definition/infox

(and Le Petit Robert only accept it in its 2020 edition, not before).
(and both of these are reference dictionaries in French French, so for the `fr_FR` language)

So, "words that are part of a language" is actually ambiguous => according to which dictionary or dictionaries?

This gets even hairier with regional differences. "armour" is a word in `en_UK` but not in `en_US`. It should be counted as a word in `en` (because the definition only cares for the language, not for the region). But then, both "armour" and "armor" should count for `en`, which requires a list of words for each of the possible subregions.

Author: I think the language input addresses this. Count up all the words you understand to be part of a particular language, that's the number of words. If the dictionary you may use for that is different from the one someone else uses, that's a change that is different because the input is different. I don't think that makes this definition ambiguous.

Reviewer: Good point. I am convinced.

The _words_ used in the following texts should be counted for the _most common language_:

> - **page title**: The value of the [document title][] of the [document][] in the [top-level browsing context][].

Reviewer: I do not understand why browsing context is needed here. "document title" doesn't seem to need it.

Author: Because the document title of iframed content is never exposed to AT (at least not as far as I'm aware).

Reviewer: Good point.
Should that be an assumption? (for the rule)
"The rule assumes that `iframe` titles are not exposed to AT and thus do not consider them"

> - **default language content**: The value of any [text nodes][] and the [accessible name][] of any [elements][] for which all of the following is true:

Reviewer: Should that also include accessible description? Or do accessible descriptions always come through text nodes (hum, probably not with the inclusion in the accessibility tree)?

Author: Good point, yeah I think so.

> - **default language content**: The value of any [text nodes][] and the [accessible name][] of any [elements][] for which all of the following is true:

Reviewer: I think there should be examples where the accessible name plays a role in determining the language.
```html
<span id="explanation" aria-hidden=true>this word has no English equivalent</span>
<span aria-labelledby="explanation">gelukkig</span>
```
or something similar.

Author: Not too keen on an aria-labelledby one, but an image makes sense. I'll work on something.

>   - The [node][] is [included in the accessibility tree][]; and

Reviewer: Why including accessible but invisible nodes and not visible but not accessible nodes?

Author: Because those nodes will not be processed by assistive technologies. The language of content in aria-hidden doesn't much matter, since it's not going to get read out.

Reviewer: Understanding 3.1.1 also list benefits for the visual parts:
> Visual browsers can display characters and scripts correctly. Media players can show captions correctly. 

And so does 3.1.2:
> This is especially important when switching between languages that read from left to right and languages that read from right to left, or when text is rendered in a language that uses a different alphabet. Users with disabilities who know all the languages used in the Web page will be better able to understand the content when each passage is rendered appropriately.

This seems to indicate that a correctly set programmatic language is important both for screen readers and visual display, hence that "visible only" words should apply.

I can imagine a situation where the visual part of the page has been translated, but the accessible hasn't. This would result in the language this definition finds being different whether we count "visible only" and "AT only" words and thus may the rule fail where it shouldn't.

---
title: Default Page Language
key: default-page-language
unambiguous: true
objective: true
input_aspects:
  - DOM tree
  - Accessibility tree
  - CSS Styling
  - Language
---

The default page language is the _most common language_ used in a [web page][]. If there are no _words_ in the page, the default language is undefined.

The _most common language_ is determined by counting the number of _words_ in the [web page][] that are part of any of the languages in the [language subtag registry][]. The same word can be part of multiple languages. If the number of words is the same, the default language is undefined.

The _words_ used in the following texts should be counted for the _most common language_:

- **page title**: The value of the [document title][] of the [document][] in the [top-level browsing context][].
- **text nodes**: The value of any [text nodes][] that is [visible][] or [included in the accessibility tree][]
- **accessible text**: The [accessible name][] and [accessible description][] of any element included in the [accessibility tree][]

> **Exception**: Words in text that comes from nodes that have an [ancestor][] in the [flat tree][] with a non-empty (`""`) `lang` attribute, except for the [root node][].

Reviewer: ```suggestion
**Exception**: Do not count words in text that comes from nodes that have an [ancestor][] in the [flat tree][] which has a non-empty (`""`) `lang` attribute and is not the [root node][].
```
The original sentence is OK-ish but I feel it has two problems: 1/ not a sentence (no verb, …) so adding the "do not count" part; 2/ "except for the root node" was separated from what it qualifies ("ancestor in the flat tree"), making it look like it apply to the full nominal group ("nodes that have an ancestor with a non-empty lang attribute").

Not super happy with what I'm suggesting either, feels very "heavy" sentence for something relatively simple…

> **Exception**: Words in text that comes from nodes that have an [ancestor][] in the [flat tree][] with a non-empty (`""`) `lang` attribute, except for the [root node][].

Reviewer: Do you mean "non-empty lang" or "valid lang"?
i.e., do you mean to exclude `<p lang="xyz">Lorem ipsum</p>`?
I am not sure how invalid lang are handled by UA/AT. If they default to the parent lang, then these should not be ignored from the count.

Author: AT are inconsistent about what they do, some default to the parent, others default to the browser's default language. That's why I think it's best to keep them out. Clearly someone meant for that to be a different language.

> **Exception**: Words in text that comes from nodes that have an [ancestor][] in the [flat tree][] with a non-empty (`""`) `lang` attribute, except for the [root node][].

Reviewer: How is language handled in a case like that:
```html
<div lang="fr"><span id="label">Hello World</span></div>
<img src="foo.jpg" aria-labelledby="label" />
```

With the current definition, "Hello World" is ignored because it has an ancestor with a lang attribute. But how is it handled by AT? Is it going to be read with the ancestor attribute or default to the page language?

Similar question for
```html
<span id="label" lang="fr">Hello World</span>
<img src="foo.jpg" aria-labelledby="label" />
```

Author: The text is included because it is part of the accessible name of the img.

Reviewer: Then, I do not understand "text that comes from nodes…" the way you seem to (i.e. that need some clarification).

In the examples I've put above, I understand that "Hello World" is text that comes from the `span` node (or more precisely from the text node which is child of the `span`). These nodes do have an ancestor with a non-empty `lang`.

From your answer, I get that you meant that "Hello World" is text that comes from the `img` node, which doesn't have an ancestor with a non-empty `lang`.

Note that the text in the `span` has actually 2 reasons to be counted: it is displayed (and announced), and it is used in an accessible name. One of these reasons is cancelled due to the `lang` attribute, but not the other. That's also a bit counter-intuitive.

The _words_ used in the following texts should be counted for the _most common language_:

- **page title**: The value of the [document title][] of the [document][] in the [top-level browsing context][].
- **text nodes**: The value of any [text nodes][] that are [visible][] or [included in the accessibility tree][]
- **accessible text**: The [accessible name][] and [accessible description][] of any element included in the [accessibility tree][]

> **Exception**: Do not count words in text that comes from nodes that have an [ancestor][] in the [flat tree][] which has a non-empty (`""`) `lang` attribute and is not the [root node][]. For accessible texts, only ancestors of the node with the accessible name is considered, the ancestry of nodes used in producing the accessible text are ignored, such as those referenced with aria-labelledby.

Reviewer: I think we should have a Passed and Failed example with an `aria-labelledby` to illustrate the second sentence.
